### PR TITLE
Pubsub new tests e2e json avro test debug

### PR DESCRIPTION
**Pull request by AnkitCLI**

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -80,7 +80,7 @@
     <google.cloud.bigtable.version>1.17.1</google.cloud.bigtable.version>
     <google.cloud.bigquery.version>1.137.1</google.cloud.bigquery.version>
     <google.cloud.kms.version>2.0.2</google.cloud.kms.version>
-    <google.cloud.pubsub.version>1.108.1</google.cloud.pubsub.version>
+    <google.cloud.pubsub.version>1.112.1</google.cloud.pubsub.version>
```
Comment (Vipinofficial11) [inline pom.xml:83]: Why are we updating the pubsub and avro version in dev dependency ?

File: pom.xml [hunk 2]
```diff
@@ -638,7 +638,7 @@
     <dependency>
       <groupId>org.apache.avro</groupId>
       <artifactId>avro-mapred</artifactId>
-      <classifier>hadoop2</classifier>
+<!--      <classifier>hadoop2</classifier>-->
```
Comment (Vipinofficial11) [inline pom.xml:641]: Why is it commented ?

File: src/e2e-test/features/pubsub/pubsubadditionalfeature/PubSubToPubSub.feature [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    And Close the Plugin Properties page
+    And Click on configure button
+    And Click on pipeline config
+    And Click on batch time and select format
+    Then Save the pipeline
+    Then Deploy the pipeline
+    Then Run the Pipeline in Runtime
+    Then Wait for pipeline to be in status: "Running" with a timeout of 240 seconds
+    Then Publish the messages
+    Then Validate OUT record count is equal to IN record count
```
Comment (Vipinofficial11) [inline src/e2e-test/features/pubsub/pubsubadditionalfeature/PubSubToPubSub.feature:51]: We are only validating the record count how about data validation.

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java [hunk 4]
```diff
@@ -75,14 +80,14 @@ public static void overrideServiceAccountFilePathIfProvided() {
...
-            }
-            return;
+        if (serviceAccountType.equalsIgnoreCase("FilePath")) {
+          PluginPropertyUtils.addPluginProp("serviceAccountType", "filePath");
+          String serviceAccountFilePath = System.getenv("SERVICE_ACCOUNT_FILE_PATH");
+          if (!(serviceAccountFilePath == null) && !serviceAccountFilePath.equalsIgnoreCase("auto-detect")) {
+            PluginPropertyUtils.addPluginProp("serviceAccount", serviceAccountFilePath);
           }
+          return;
+        }
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: The indentation was changed here but the code changes are unrelated to the PR, please revert the changes so its not highlighted. (Review all the changes for this)

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java [hunk 5]
```diff
@@ -480,12 +485,90 @@ private static String createGCSBucketWithFilesAndFolder(String folderPath) throw
...
+
+  @After(order = 1, value = "@PUBSUB_SOURCE_TEST")
+  public static void deleteSourcePubSubTopic() {
+    try {
+      PubSubClient.deleteTopic(pubSubSourceTopic);
+      BeforeActions.scenario.write("Deleted target PubSub topic " + pubSubSourceTopic);
+      pubSubSourceTopic = StringUtils.EMPTY;
+    } catch (Exception e) {
+      if (e.getMessage().contains("Invalid resource name given") || e.getMessage().contains("Resource not found")) {
+
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: The If block is empty please add the necessary code/logic.

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java [hunk 6]
```diff
@@ -480,12 +485,90 @@ private static String createGCSBucketWithFilesAndFolder(String folderPath) throw
...
+  @After(order = 2, value = "@PUBSUB_SCHEMA_TOPIC_TEST")
+  public static void deleteSourcePubSubSchemaTopic() {
+    try {
+      PubSubClient.deleteTopic(pubSubSourceTopic);
+      BeforeActions.scenario.write("Deleted target PubSub topic " + pubSubSourceTopic);
+      pubSubSourceTopic = StringUtils.EMPTY;
+    } catch (Exception e) {
+      if (e.getMessage().contains("Invalid resource name given") || e.getMessage().contains("Resource not found")) {
+
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: empty block

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java [hunk 7]
```diff
@@ -480,12 +485,90 @@ private static String createGCSBucketWithFilesAndFolder(String folderPath) throw
...
   @Before(order = 1, value = "@PUBSUB_SINK_TEST")
   public static void createTargetPubSubTopic() {
     pubSubTargetTopic = "cdf-e2e-test-" + UUID.randomUUID();
     BeforeActions.scenario.write("Target PubSub topic " + pubSubTargetTopic);
   }
 
+  @After(order = 1, value = "@PUBSUB_SCHEMA_TEST")
+  public static void deletePubSubSchema() throws IOException {
+    PubSubClient.deleteSchema(PluginPropertyUtils.pluginProp(ConstantsUtil.PROJECT_ID), pubSubSchemaId);
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: You are using ```PluginPropertyUtils.pluginProp(ConstantsUtil.PROJECT_ID)``` multiple times in the code make a global variable for this and reuse it.

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java [hunk 8]
```diff
@@ -790,7 +873,7 @@ public static void deleteBQTableForBQExecuteTest() throws IOException, Interrupt
       } else {
         Assert.fail(e.getMessage());
       }
-  }
+    }
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java:879]: revert

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java [hunk 9]
```diff
@@ -883,7 +966,7 @@ public static void setUpdateQueryBackWithTableDetailsPlaceholder() {
 
   @Before(order = 1, value = "@BQ_INSERT_SOURCE_TEST")
   public static void createSourceBQInsertTable() throws IOException, InterruptedException {
-    bqSourceTable = "E2E_SOURCE_" + UUID.randomUUID().toString().replaceAll("-" , "_");
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: revert

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java [hunk 10]
```diff
@@ -1028,8 +1112,8 @@ public static void createSinkBQExistingDatatypeTable() throws IOException, Inter
       BeforeActions.scenario.write("Error inserting the record in the table" + e.getStackTrace());
     }
 
-    PluginPropertyUtils.addPluginProp(" bqTargetTable",  bqTargetTable);
-    BeforeActions.scenario.write("BQ Target Table " +  bqTargetTable + " updated successfully");
+    PluginPropertyUtils.addPluginProp(" bqTargetTable", bqTargetTable);
+    BeforeActions.scenario.write("BQ Target Table " + bqTargetTable + " updated successfully");
   }
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: Revert all the above changes which are highlighted because of indentation.

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java [hunk 11]
```diff
@@ -1286,6 +1370,7 @@ public static void createSinkBQInsertTable() throws IOException, InterruptedExce
     PluginPropertyUtils.addPluginProp(" bqTargetTable", bqTargetTable);
     BeforeActions.scenario.write("BQ Target Table " + bqTargetTable + " updated successfully");
   }
+
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java:1376]: remove extra line

File: src/e2e-test/java/io/cdap/plugin/pubsub/actions/PubSubActions.java [hunk 12]
```diff
@@ -85,4 +89,9 @@ public static void enterSubscription(String subscription) {
   public static void enterNumberOfReaders(String numberOfReaders) {
     ElementHelper.replaceElementValue(PubSubLocators.numberOfReaders, numberOfReaders);
   }
+
+  public static void selectDataType() {
+    Select select = new Select(PubSubLocators.messageDataType);
+    select.selectByIndex(9);
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/pubsub/actions/PubSubActions.java:95]: Why are we hardcoding the index here are we sure the desired message data type will be at 9th index at all occurrences ?

File: src/e2e-test/java/io/cdap/plugin/pubsub/locators/PubSubLocators.java [hunk 13]
```diff
@@ -64,6 +64,20 @@ public class PubSubLocators {
 
   @FindBy(how = How.XPATH, using = "//input[@data-cy='numberOfReaders']")
   public static WebElement numberOfReaders;
+  @FindBy(how = How.XPATH, using = "//input[@placeholder='Field name' and @value='']")
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/pubsub/locators/PubSubLocators.java:67]: If the locator doesn't have a data-cy or data-test-id  try writing the locator which won't break. (Also no need to include value here as an and)

File: src/e2e-test/java/io/cdap/plugin/pubsub/locators/PubSubLocators.java [hunk 14]
```diff
@@ -64,6 +64,20 @@ public class PubSubLocators {
...
+  public static WebElement addField;
+  @FindBy(how = How.XPATH, using = "//span[contains(text(), \"Batch interval\")]//following-sibling::div//select[1]")
+  public static WebElement batchTime;
+  @FindBy(how = How.XPATH, using = "//span[contains(text(), \"Batch interval\")]//following-sibling::div//select[2]")
+  public static WebElement timeSelect;
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/pubsub/locators/PubSubLocators.java:72]: Try writing better locators for time select and batch time if possible.

File: src/e2e-test/java/io/cdap/plugin/pubsub/locators/PubSubLocators.java [hunk 15]
```diff
@@ -64,6 +64,20 @@ public class PubSubLocators {
...
+  @FindBy(how = How.XPATH, using = "//button[@data-testid='config-apply-close']")
+  public static WebElement saveButton;
+  @FindBy(how = How.XPATH, using = "//*[@data-cy='pipeline-configure-modeless-btn']")
+  public static WebElement configButton;
+  @FindBy(how = How.XPATH, using = "//*[@data-cy='tab-content-Pipeline config']")
+  public static WebElement pipelineConfig;
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/pubsub/locators/PubSubLocators.java:78]: Why use * instead of tag

File: src/e2e-test/java/io/cdap/plugin/pubsub/stepsdesign/PubSubSink.java [hunk 16]
```diff
@@ -234,4 +237,33 @@ public void enterRuntimeArgumentValueForPubSubSinkPropertyTopicKey(String runtim
   public void clickOnPreviewDataForPubSubSink() {
     openSinkPluginPreviewData("GooglePublisher");
   }
+
+  @Then("Subscribe to the messages")
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/pubsub/stepsdesign/PubSubSink.java]: Subscribe to the Pub/Sub messages seems more detailed and informative

File: src/e2e-test/java/io/cdap/plugin/pubsub/stepsdesign/PubSubSink.java [hunk 17]
```diff
@@ -234,4 +237,33 @@ public void enterRuntimeArgumentValueForPubSubSinkPropertyTopicKey(String runtim
...
+  public void subscribeToTheMessages() throws InterruptedException {
+    TimeUnit time = TimeUnit.SECONDS;
+    time.sleep(60);
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/pubsub/stepsdesign/PubSubSink.java:244]: Why are waiting here

File: src/e2e-test/java/io/cdap/plugin/pubsub/stepsdesign/PubSubSink.java [hunk 18]
```diff
@@ -234,4 +237,33 @@ public void enterRuntimeArgumentValueForPubSubSinkPropertyTopicKey(String runtim
...
+    PubSubClient.subscribeAsyncExample(PluginPropertyUtils.pluginProp(ConstantsUtil.PROJECT_ID),
+                                       TestSetupHooks.pubSubSourceSubscription);
+  }
+
+
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/pubsub/stepsdesign/PubSubSink.java]: remove extra space

File: src/e2e-test/java/io/cdap/plugin/pubsub/stepsdesign/PubSubSink.java [hunk 19]
```diff
@@ -234,4 +237,33 @@ public void enterRuntimeArgumentValueForPubSubSinkPropertyTopicKey(String runtim
...
+  }
+
+  @And("Click on configure button")
+  public void clickOnConfigureButton() {
+    ElementHelper.clickOnElement(PubSubLocators.configButton);
+  }
+
+  @And("Click on pipeline config")
+  public void clickOnPipelineConfig() {
+    ElementHelper.clickOnElement(PubSubLocators.pipelineConfig);
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/pubsub/stepsdesign/PubSubSink.java:265]: For these click steps can't we use the common steps from framework

File: src/e2e-test/java/io/cdap/plugin/pubsub/stepsdesign/PubSubSink.java [hunk 20]
```diff
@@ -234,4 +237,33 @@ public void enterRuntimeArgumentValueForPubSubSinkPropertyTopicKey(String runtim
...
+  }
+
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/pubsub/stepsdesign/PubSubSink.java]: extra space

File: src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java [hunk 21]
```diff
@@ -36,6 +68,22 @@ public static Topic createTopic(String topicId) throws IOException {
     }
   }
 
+  // Create the subscription
+  public static void createSubscription(String subscriptionId, String topicId) throws IOException {
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java:83]: Add a proper java doc instead of a one liner comment.

File: src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java [hunk 22]
```diff
@@ -36,6 +68,22 @@ public static Topic createTopic(String topicId) throws IOException {
...
+    try (SubscriptionAdminClient subscriptionAdminClient = SubscriptionAdminClient
+      .create(SubscriptionAdminSettings.newBuilder().build())) {
+      TopicName topicName = TopicName.of(PluginPropertyUtils.pluginProp(ConstantsUtil.PROJECT_ID), topicId);
+      subscriptionName = ProjectSubscriptionName.of(PluginPropertyUtils.
+                                                      pluginProp(ConstantsUtil.PROJECT_ID), subscriptionId);
+      subscriptionAdminClient.createSubscription(subscriptionName, topicName, PushConfig.getDefaultInstance(), 60);
+      System.out.println("Subscription created: " + subscriptionName.toString());
+    } catch (AlreadyExistsException e) {
+      assert subscriptionName != null;
+      System.out.println("Subscription already exists: " + subscriptionName);
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java]: Remove system.out and use Before or After Actions.

File: src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java [hunk 23]
```diff
@@ -36,6 +68,22 @@ public static Topic createTopic(String topicId) throws IOException {
...
+    ProjectSubscriptionName subscriptionName = null;
+    try (SubscriptionAdminClient subscriptionAdminClient = SubscriptionAdminClient
+      .create(SubscriptionAdminSettings.newBuilder().build())) {
+      TopicName topicName = TopicName.of(PluginPropertyUtils.pluginProp(ConstantsUtil.PROJECT_ID), topicId);
+      subscriptionName = ProjectSubscriptionName.of(PluginPropertyUtils.
+                                                      pluginProp(ConstantsUtil.PROJECT_ID), subscriptionId);
+      subscriptionAdminClient.createSubscription(subscriptionName, topicName, PushConfig.getDefaultInstance(), 60);
+      System.out.println("Subscription created: " + subscriptionName.toString());
+    } catch (AlreadyExistsException e) {
+      assert subscriptionName != null;
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java]: Assert properly with messages

File: src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java [hunk 24]
```diff
@@ -36,6 +68,22 @@ public static Topic createTopic(String topicId) throws IOException {
...
+  // Create the subscription
+  public static void createSubscription(String subscriptionId, String topicId) throws IOException {
+    ProjectSubscriptionName subscriptionName = null;
+    try (SubscriptionAdminClient subscriptionAdminClient = SubscriptionAdminClient
+      .create(SubscriptionAdminSettings.newBuilder().build())) {
+      TopicName topicName = TopicName.of(PluginPropertyUtils.pluginProp(ConstantsUtil.PROJECT_ID), topicId);
+      subscriptionName = ProjectSubscriptionName.of(PluginPropertyUtils.
+                                                      pluginProp(ConstantsUtil.PROJECT_ID), subscriptionId);
+      subscriptionAdminClient.createSubscription(subscriptionName, topicName, PushConfig.getDefaultInstance(), 60);
+      System.out.println("Subscription created: " + subscriptionName.toString());
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java]: remove the sout

File: src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java [hunk 25]
```diff
@@ -56,4 +104,186 @@ public static String getTopicCmekKey(String topicId) throws IOException {
     return getTopic(topicId).getKmsKeyName();
   }
 
+  public static void publishWithErrorHandlerExample(String projectId, String topicId, List<String> jsonMessages)
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java]: Add a javadoc

File: src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java [hunk 26]
```diff
@@ -56,4 +104,186 @@ public static String getTopicCmekKey(String topicId) throws IOException {
...
+    throws IOException, InterruptedException {
+    TopicName topicName = TopicName.of(PluginPropertyUtils.pluginProp(ConstantsUtil.PROJECT_ID), topicId);
+    Publisher publisher = null;
+    try {
+      publisher = Publisher.newBuilder(topicName).build();
+
+      for (final String message : jsonMessages) {
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java]: change the jsonMessage to something relevant

File: src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java [hunk 27]
```diff
@@ -56,4 +104,186 @@ public static String getTopicCmekKey(String topicId) throws IOException {
...
+        // Adding an asynchronous callback to handle success / failure
+        ApiFutures.addCallback(future, new ApiFutureCallback<String>() {
+
+          @Override
+          public void onFailure(Throwable throwable) {
+            if (throwable instanceof ApiException) {
+              ApiException apiException = ((ApiException) throwable);
+              // details on the API exception
+              System.out.println(apiException.getStatusCode().getCode());
+              System.out.println(apiException.isRetryable());
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java]: remove system.out.println

File: src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java [hunk 28]
```diff
@@ -56,4 +104,186 @@ public static String getTopicCmekKey(String topicId) throws IOException {
...
+  public static void deleteSubscription(String projectId, String subscriptionId) throws IOException {
+    try (SubscriptionAdminClient subscriptionAdminClient = SubscriptionAdminClient.create()) {
+      ProjectSubscriptionName subscriptionName = ProjectSubscriptionName.of(projectId, subscriptionId);
+      try {
+        subscriptionAdminClient.deleteSubscription(subscriptionName);
+      } catch (NotFoundException e) {
+        System.out.println(e.getMessage());
+      }
+    }
+  }
```
Comment (Vipinofficial11) [inline src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java:291]: Fix the code indentation and remove all the system.out.println

File: src/e2e-test/resources/pluginParameters.properties [hunk 29]
```diff
@@ -181,6 +181,11 @@ bqUpdateTableSchemaTrue=True
 clusterValue=transaction_date
 TableKey=PersonID
 bqSourceTable=dummy
+#pubSubSourceSubscription=dummy
+#pubSubSourceTopic=dummy
```
Comment (Vipinofficial11) [inline src/e2e-test/resources/pluginParameters.properties]: remove comments if not required

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQueryWindowsAggregationSQLBuilder.java [hunk 30]
```diff
@@ -21,7 +21,6 @@
 import io.cdap.cdap.etl.api.relational.Expression;
 import io.cdap.plugin.gcp.bigquery.relational.SQLExpression;
 import io.cdap.plugin.gcp.bigquery.sqlengine.util.BigQuerySQLEngineUtils;
-import org.mortbay.log.Log;
```
Comment (Vipinofficial11) [inline src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQueryWindowsAggregationSQLBuilder.java:24]: dev change, please revert.

File: pom.xml (hunk 1 above)
Comment (AnkitCLI) [reply pom.xml:83]: We need this is the old version and it is not compatible with various components in the code.

File: pom.xml (hunk 2 above)
Comment (AnkitCLI) [reply pom.xml:641]: due to the version issue it is commented.

File: src/e2e-test/features/pubsub/pubsubadditionalfeature/PubSubToPubSub.feature (hunk 3 above)
Comment (AnkitCLI) [reply src/e2e-test/features/pubsub/pubsubadditionalfeature/PubSubToPubSub.feature:51]: we are not going with data validation right now.

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java (hunk 5 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: added

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java (hunk 6 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: added.

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java [hunk 31]
```diff
@@ -984,29 +1072,29 @@ public static void createSinkBQExistingTable() throws IOException, InterruptedEx
...
-                                                              "'2021-01-28'," + "0.0904809091," +
-                                                              "'2018-03-10 04:50:01 UTC' " +
-                                                              ") ");
-      } catch (NoSuchElementException e) {
-        // Insert query does not return any record.
-        // Iterator on TableResult values in getSoleQueryResult method throws NoSuchElementException
-        BeforeActions.scenario.write("Error inserting the record in the table" + e.getStackTrace());
-      }
-      PluginPropertyUtils.addPluginProp("bqSourceTable", bqSourceTable);
-      BeforeActions.scenario.write("BQ Source Table " + bqSourceTable + " created successfully");
```
Comment (bharatgulati) [inline src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java:1008]: revert

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java (hunk 7 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: directly used from pluginparameters file.

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java (hunk 9 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: done.

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java (hunk 10 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java]: done.

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java [hunk 32]
```diff
@@ -1295,5 +1384,6 @@ private static String createGCSBucketLifeCycle() throws IOException, URISyntaxEx
   @Before(order = 1, value = "@GCS_SINK_MULTI_PART_UPLOAD")
   public static void createBucketWithLifeCycle() throws IOException, URISyntaxException {
     gcsTargetBucketName = createGCSBucketLifeCycle();
-    BeforeActions.scenario.write("GCS target bucket name - " + gcsTargetBucketName); }
```
Comment (bharatgulati) [inline src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java:1298]: revert

File: src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java (hunk 11 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/common/stepsdesign/TestSetupHooks.java:1376]: removed.

File: src/e2e-test/java/io/cdap/plugin/pubsub/actions/PubSubActions.java (hunk 12 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/pubsub/actions/PubSubActions.java:95]: yes.

File: src/e2e-test/java/io/cdap/plugin/pubsub/locators/PubSubLocators.java (hunk 13 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/pubsub/locators/PubSubLocators.java:67]: this is the complete locator as we are adding the fields in which this value section is empty.

File: src/e2e-test/java/io/cdap/plugin/utils/State.java [hunk 33]
```diff
@@ -0,0 +1,419 @@
...
+import org.apache.avro.specific.SpecificData;
+import org.apache.avro.util.Utf8;
+
+/** A list of states in the United States of America. */
+@org.apache.avro.specific.AvroGenerated
+public class State extends org.apache.avro.specific.SpecificRecordBase
+  implements org.apache.avro.specific.SpecificRecord {
+  public static final org.apache.avro.Schema SCHEMA = new org.apache.avro.Schema.Parser()
+    .parse("{\"type\":\"record\",\"name\":\"State\",\"namespace\":\"utilities\",\"doc\":\"A list of states in the" +
+             " United States of America.\",\"fields\":[{\"name\":\"name\",\"type\":\"string\",\"doc\":\"The common " +
```
Comment (bharatgulati) [inline src/e2e-test/java/io/cdap/plugin/utils/State.java:20]: store the schema in the resource file and call it from there.

File: src/e2e-test/java/io/cdap/plugin/pubsub/locators/PubSubLocators.java (hunk 14 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/pubsub/locators/PubSubLocators.java:72]: These locators are referenced from salesforce runtime code.

File: src/e2e-test/java/io/cdap/plugin/pubsub/locators/PubSubLocators.java (hunk 15 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/pubsub/locators/PubSubLocators.java:78]: same comment as above.

File: src/e2e-test/java/io/cdap/plugin/pubsub/stepsdesign/PubSubSink.java (hunk 16 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/pubsub/stepsdesign/PubSubSink.java]: done.

File: src/e2e-test/java/io/cdap/plugin/pubsub/stepsdesign/PubSubSink.java (hunk 17 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/pubsub/stepsdesign/PubSubSink.java:244]: we are waiting here as we are using pubsub realtime so we need to wait for sometime to reflect the messages to subscribe it.

File: src/e2e-test/java/io/cdap/plugin/utils/State.java [hunk 34]
```diff
@@ -0,0 +1,419 @@
...
+   */
+  public State() {
+  }
+
+  /**
+   * All-args constructor.
+   * @param name The common name of the state.
+   * @param postabbr The postal code abbreviation of the state.
+   */
+  public State(CharSequence name, CharSequence postabbr) {
```
Comment (bharatgulati) [inline src/e2e-test/java/io/cdap/plugin/utils/State.java:53]: use naming convention for variable CharSequence correctly. Change it to something more relevant.

File: src/e2e-test/java/io/cdap/plugin/pubsub/stepsdesign/PubSubSink.java (hunk 18 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/pubsub/stepsdesign/PubSubSink.java]: removed.

File: src/e2e-test/java/io/cdap/plugin/pubsub/stepsdesign/PubSubSink.java (hunk 19 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/pubsub/stepsdesign/PubSubSink.java:265]: no we cannot use it from the common steps as it is not included in framework right now.

File: src/e2e-test/java/io/cdap/plugin/utils/State.java [hunk 35]
```diff
@@ -0,0 +1,419 @@
+/**
```
Comment (bharatgulati) [inline src/e2e-test/java/io/cdap/plugin/utils/State.java:1]: is this dev code?

File: src/e2e-test/java/io/cdap/plugin/pubsub/stepsdesign/PubSubSink.java (hunk 20 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/pubsub/stepsdesign/PubSubSink.java]: removed

File: src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java (hunk 21 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java:83]: added a java doc.

File: src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java (hunk 22 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java]: added looger.info

File: src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java (hunk 23 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java]: added.

File: src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java (hunk 24 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java]: removed.

File: src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java (hunk 25 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java]: added.

File: src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java (hunk 26 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java]: changed to messages.

File: src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java (hunk 27 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java]: removed.

File: src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java (hunk 28 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/utils/PubSubClient.java:291]: done.

File: src/e2e-test/resources/pluginParameters.properties (hunk 29 above)
Comment (AnkitCLI) [reply src/e2e-test/resources/pluginParameters.properties]: removed.

File: src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQueryWindowsAggregationSQLBuilder.java (hunk 30 above)
Comment (AnkitCLI) [reply src/main/java/io/cdap/plugin/gcp/bigquery/sqlengine/builder/BigQueryWindowsAggregationSQLBuilder.java:24]: removd.

File: src/e2e-test/java/io/cdap/plugin/utils/State.java (hunk 35 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/utils/State.java:1]: same comment as above.

File: src/e2e-test/java/io/cdap/plugin/utils/State.java (hunk 34 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/utils/State.java:53]: same comment as above.

File: src/e2e-test/java/io/cdap/plugin/utils/State.java (hunk 33 above)
Comment (AnkitCLI) [reply src/e2e-test/java/io/cdap/plugin/utils/State.java:20]: this is the autogenerated file from avro tools.
